plugins/router/f5/f5.go

```diff
@@ -0,0 +1,1743 @@
+	// we always check whether it is initialised before using it, but
+	// initialising it to an empty map here saves a REST call later the first
+	// time f5.PoolHasMember is invoked with poolname.
+	f5.poolMembers[poolname] = map[string]bool{}
+
+	glog.V(4).Infof("Pool %s created.", poolname)
+
+	return nil
+}
+
+// Delete the pool named poolname from F5, and delete
+// f5.poolMembers[poolname].
+func (f5 *f5LTM) DeletePool(poolname string) error {
+	url := fmt.Sprintf("https://%s/mgmt/tm/ltm/pool/%s", f5.host, poolname)
+
+	err := f5.delete(url, nil)
+	if err != nil {
+		return err
+	}
+
+	// Note: We *must* use delete here rather than merely assigning false because
+	// len() includes false items, and we want len() to return an accurate count
+	// of members.  Also, we probably save some memory by using delete.
+	delete(f5.poolMembers, poolname)
+
+	glog.V(4).Infof("Pool %s deleted.", poolname)
+
+	return nil
+}
+
+// Return f5.poolMembers[poolname], first initializing it from F5 if it is zero.
+func (f5 *f5LTM) GetPoolMembers(poolname string) (map[string]bool, error) {
+	members, ok := f5.poolMembers[poolname]
+	if ok {
+		return members, nil
+	}
+
+	url := fmt.Sprintf("https://%s/mgmt/tm/ltm/pool/%s/members",
+		f5.host, poolname)
+
+	res := f5Pool{}
+
+	err := f5.get(url, &res)
+	if err != nil {
+		return nil, err
+	}
+
+	// Note that we do not initialise f5.poolMembers[poolname] unless we know that
+	// the pool exists (i.e., the above GET request for the pool succeeds).
+	f5.poolMembers[poolname] = map[string]bool{}
+
+	for _, member := range res.Members {
+		f5.poolMembers[poolname][member.Name] = true
+	}
+
+	return f5.poolMembers[poolname], nil
+}
+
+// Check whether the specified pool exists.  Internally, this method uses
+// f5.poolMembers[poolname], as a side effect initialising it if it is zero.
+func (f5 *f5LTM) PoolExists(poolname string) (bool, error) {
+	_, err := f5.GetPoolMembers(poolname)
+	if err == nil {
+		return true, nil
+	}
+
+	if err.(F5Error).httpStatusCode == 404 {
+		return false, nil
+	}
+
+	return false, err
+}
+
+// Check whether the given member is in the pool named poolname on F5.
+// Internally, this method uses f5.poolMembers[poolname], initialising
+// it first if it is zero.
+func (f5 *f5LTM) PoolHasMember(poolname, member string) (bool, error) {
+	members, err := f5.GetPoolMembers(poolname)
+	if err != nil {
+		return false, err
+	}
+
+	return members[member], nil
+}
+
+// Add member to the pool named poolname on F5, and update
+// f5.poolMembers[poolname].
+func (f5 *f5LTM) AddPoolMember(poolname, member string) error {
+	hasMember, err := f5.PoolHasMember(poolname, member)
+	if err != nil {
+		return err
+	}
+	if hasMember {
+		glog.V(4).Infof("Pool %s already has member %s.\n", poolname, member)
+		return nil
+	}
+
+	glog.V(4).Infof("Adding pool member %s to pool %s.", member, poolname)
+
+	url := fmt.Sprintf("https://%s/mgmt/tm/ltm/pool/%s/members",
```

Comment:
Just something like a `map[string]string`? That might obfuscate more than improve clarity.  Or maybe something fancier like `map[string]&getResource` where `getResource` would be a function that checked its arguments? That feels slightly over-engineered, but it might e slightly safer.